### DISCUSSION TITLE
Why not affect "updateNodeInternal()" after calling "updateNodeData()"?

### DISCUSSION THREAD

**Discussion by peanudge** [xyflow/xyflow#4613] [Q&A] [answered]
I met strange phenomenon by timing. Please let me know advice if i missed something.

```tsx

const ComponentA = () => {

   const { updateNodeData, getNode } = useReactFlow<
        CustomNodeType,
        CustomEdgeType
    >()
    const updateNodeInternal = useUpdateNodeInternals()
    const handleClick = () => {
        updateNodeData("node-id", (node) => return {...node.data, handleCount: 2 });
        
       // updateNodeInternal("node-id"); // This not affect handle on graph ui
       setTimeout(() => { updateNodeInternal("node-id"); }, 300); // Work!
    }
   return <></>
}
```

Can you guide me? why `ReactFlowInstance.updateNodeData()` or `ReactFlowInstance.updateNode()` require short interval time before `updateNodeInternal()` for updating **handle** positions?

**Comment by bcakmakoglu** [comment, marked as answer]
Assuming that `handleCount` changes the amount of `<Handle>` components that are rendered in your custom node you have to consider the time it takes for React to react to the change in your `handleCount`, render the `<Handle>` components and only after they have been mounted and rendered would `updateNodeInternals` have any effect, otherwise there are no `<Handle>` components that could be re-measured.

**Reply by peanudge** [reply]
> only after they have been mounted and rendered would updateNodeInternals have any effect

Thank you for helpful above information.

**Comment by moklick** [comment]
`updateNodeData` is not a synchronous function. It triggers an update, that needs to wait for the next render to happen. You could handle the update in the custom node like this:

```js 
useEffect(() => {
  updateNodeInternals(id);
}, [handleCount])
```

**Reply by peanudge** [reply]
Thanks! moklick

Just curious.
+) Only `update__`  functions  of `ReactFlowInstance` are not synchronous functions? How about `set__`, `delete___`  ? Same thing?

**Reply by moklick** [reply]
Everything that alters nodes/edges is async.

**Comment by peanudge** [comment]
I decided below solution! thank you for sharing advices.

```tsx
   const updateNodeInternal = useUpdateNodeInternals()
   const selectedNode = useNodesData<CustomNodeType>(selectedNodeId ?? '')
   
    useEffect(() => {
        if (selectedNode !== null) {
            updateNodeInternal(selectedNode.id)
        }
    }, [selectedNode, updateNodeInternal])
```
